L1Trigger/L1TTrackMatch/plugins/TPFastJetProducer.cc

```diff
@@ -0,0 +1,222 @@
+      if (MCTruthTTStubHandle->findTrackingParticlePtr(theStubRef).isNull() && hasStubInLayer[layer] < 2)
+        hasStubInLayer[layer] = 1;
+      else
+        hasStubInLayer[layer] = 2;
+    }
+
+    int nStubLayerTP = 0;
+    int nStubLayerTP_g = 0;
+    for (int isum : hasStubInLayer) {
+      if (isum >= 1)
+        nStubLayerTP += 1;
+      if (isum == 2)
+        nStubLayerTP_g += 1;
+    }
+
+    // tp quality cuts to match L1 tracks
+    if (tp_pt < tpPtMin_)
+      continue;
+    if (fabs(tp_eta) > tpEtaMax_)
+      continue;
+    if (nStubTP < tpNStubMin_)
+      continue;
+    if (nStubLayerTP < tpNStubLayerMin_)
+      continue;
+    if (fabs(tp_z0) > tpZMax_)
+      continue;
+    if (tp_charge == 0.)  // extra check that all tps are charged
+      continue;
+    if (tp_eventid > 0)  // only select hard interaction tps
+      continue;
+
+    fastjet::PseudoJet psuedoJet(iterTP->px(), iterTP->py(), iterTP->pz(), iterTP->energy());
+    JetInputs.push_back(psuedoJet);                // input tps for clustering
+    JetInputs.back().set_user_index(this_tp - 1);  // save tp index in the collection
+  }                                                // end loop over tps
+
+  fastjet::ClusterSequence cs(JetInputs, jet_def);  // define the output jet collection
+  std::vector<fastjet::PseudoJet> JetOutputs =
+      fastjet::sorted_by_pt(cs.inclusive_jets(0));  // output jet collection, pT-ordered
+
+  for (unsigned int ijet = 0; ijet < JetOutputs.size(); ++ijet) {
+    math::XYZTLorentzVector jetP4(
+        JetOutputs[ijet].px(), JetOutputs[ijet].py(), JetOutputs[ijet].pz(), JetOutputs[ijet].modp());
+    float sumpt = 0;
+    float avgZ = 0;
+    std::vector<edm::Ptr<TrackingParticle>> tpPtrs;
+    std::vector<fastjet::PseudoJet> fjConstituents = fastjet::sorted_by_pt(cs.constituents(JetOutputs[ijet]));
+
+    for (unsigned int i = 0; i < fjConstituents.size(); ++i) {
+      auto index = fjConstituents[i].user_index();
+      edm::Ptr<TrackingParticle> tpPtr(TrackingParticleHandle, index);
+      tpPtrs.push_back(tpPtr);  // tracking particles in the jet
+      sumpt = sumpt + tpPtr->pt();
+      avgZ = avgZ + tpPtr->pt() * tpPtr->z0();
+    }
+    avgZ = avgZ / sumpt;
+    edm::Ref<JetBxCollection> jetRef;
+    std::vector<edm::Ptr<TTTrack<Ref_Phase2TrackerDigi_>>> dummyL1TrackPtrs;  // can't create TkJet with tp references
+    TkJet tpJet(jetP4, dummyL1TrackPtrs, avgZ, fjConstituents.size(), 0, 0, 0, false);
+    TPFastJets->push_back(tpJet);
+  }  //end loop over Jet Outputs
+
+  iEvent.put(std::move(TPFastJets), "TPFastJets");
+}
+
+void TPFastJetProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  //The following says we do not know what parameters are allowed so do no validation
+  // Please change this to state exactly what you do use, even if it is no parameters
+  edm::ParameterSetDescription desc;
+  desc.setUnknown();
+  descriptions.addDefault(desc);
+}
```

Comment:
Please add a default description for the parameters in your module.

Reply:
Done.